Multi-level navigation

pdmosses opened the pull request:
This PR is intended to supersede #192, and incorporates #448. It aims to fix #219, #447, and to respond to the discussion in #425 and in the comments on #192.

A summary of the background for this PR, and an overview of the code, are available [here](https://github.com/pdmosses/just-the-docs/blob/rec-nav-2/_tests/navigation/recursion/about.md).

The [`rec-nav-2`](https://github.com/pdmosses/just-the-docs/tree/rec-nav-2) branch used for this PR is based on ~~the latest~~ an outdated release of the theme ([v0.3.3](https://github.com/just-the-docs/just-the-docs/releases/tag/v0.3.3), October 2020).

To test `rec-nav-2` locally as a remote theme:
- Specify in  `_config.yml`:
  ```
  remote_theme: pdmosses/just-the-docs@rec-nav-2
  ```
- Specify in `Gemfile`:
  ```
  gem "just-the-docs", github: "pdmosses/just-the-docs", branch: "rec-nav-2"
  ```
- Run:
  ```
  bundle update
  bundle exec jekyll serve
  ```

The proposed updated documentation pages for the theme navigation structure are available [here](https://pdmosses.github.io/just-the-docs/docs/navigation-structure/).

To serve the `rec-nav-2` documentation pages  locally and display the results of the regression tests:
- Download the [`rec-nav-2`](https://github.com/pdmosses/just-the-docs/tree/rec-nav-2) branch
- Run:
  ```
  bundle update
  bundle exec jekyll serve --config _config.yml,_tests.yml
  ```

pdmosses commented:
The last three commits above remove the need for avoiding a special character (`^`) in titles, and add regression tests where a page title is a prefix or suffix of a potential ancestor title.

jtebert commented:
From a quick test, I believe that this also fixes the issue I opened yesterday (#492) about incorrect breadcrumbs and table of contents in collections.

pdmosses commented:
@jtebert thanks for testing this PR with multiple collections – I will add them to the regression tests.

pdmosses commented:
N.B. If further collections are added, the TOC/breadcrumbs links in the `Tests` collection will be incorrect, as reported in #492.

pyorot commented:
If I rocket-react then hopefully this gets reviewed sooner :D
I'd be interested to know the roadmap for the project, what the priorities are, and if it's still active; maybe I should @ the maintainer?
Also interested in if there's an example website based on this PR, and how ready for use it is. I was hoping it'd be merged before I needed it, but it's been 5 months since the last commit.

pdmosses commented:
@pyorot 

> Also interested in if there's an example website based on this PR, and how ready for use it is

I've tested the `rec-nav-2` branch used for this PR on two websites:

- https://pdmosses.github.io/just-the-docs/ (built together with regression tests)
- https://plancomps.github.io/CBS-beta/ (using the above as a remote theme)

The latter site has about 150 pages, and takes about one minute to build (on a laptop). The navigation on both sites seemed to work as intended; I believe it to be backwards compatible with v0.3.3. @jtebert has tried using it, but otherwise I haven't had any feedback. Good to hear that you're interested!

PR #578 combines several of my other current PRs. The branch [combination-rec-nav](https://github.com/pdmosses/just-the-docs/tree/combination-rec-nav) adds the multi-level navigation implementation from the present PR to the branch used for #578. It includes updated documentation for the navigation structure, and reorganised and extended navigation tests.

Both the above websites are now based on the [combination-rec-nav](https://github.com/pdmosses/just-the-docs/tree/combination-rec-nav) branch. I've also added support for sibling links to that branch, adapting the ideas of #394 to multi-level navigation.

 I haven't yet submitted a new PR for my new branch. I would expect the next release to be a minor one, focusing more on fixing bugs than on adding new features. However, I don't know of any roadmap for future releases: my role as collaborator is summarised [here](https://github.com/pmarsceill/just-the-docs/discussions/304#discussioncomment-6528).

If you'd like to try multi-level navigation in the meantime, you could use [combination-rec-nav](https://github.com/pdmosses/just-the-docs/tree/combination-rec-nav) (or a copy of it) as a remote theme. Please follow up here in case of any difficulties that arise.

droy2017 commented:
Hello, I am creating a technical document using Jekyll-Just The Docs, where we need multi-level navigation.  We already have the 70% of the site built, so how can we add this milt-navigation feature as it is in a different branch of just the docs. I tried following below steps as recommended above, but it did not work.

1) Updated _config.yaml with the remote theme
2) Updated the gemfile 
3) Executed "bundle exec Jekyll serve"

I have attached couple of snapshots. 
<img width="777" alt="jtd-snapshot02" src="https://user-images.githubusercontent.com/29459758/111283654-5b820a80-8665-11eb-9d9f-501f3da0cc7c.PNG">
<img width="636" alt="jtd-snapshot01" src="https://user-images.githubusercontent.com/29459758/111283657-5c1aa100-8665-11eb-96f7-9f2120412d6c.PNG">

pdmosses commented:
@droy2017 you wrote:

>  I tried following below steps as recommended above, but it did not work.

Could you be more specific? Did jekyll fail to build the site locally, or was the site built but not displaying correctly?

> 1. Updated _config.yaml with the remote theme
> 2. Updated the gemfile
> 3. Executed "bundle exec Jekyll serve"

Did you also do `bundle update`? That is always needed after changing the gemfile.

Try using the [gemfile from CBS-beta](https://github.com/plancomps/CBS-beta/blob/master/Gemfile), which is using this branch.

If you intend to use the dark color scheme, I recommend using the [combination-rec-nav](https://github.com/pdmosses/just-the-docs/tree/combination-rec-nav) branch instead of rec-nav-2.

droy2017 commented:
> 
> 
> @droy2017 you wrote:
> 
> > I tried following below steps as recommended above, but it did not work.
> 
> Could you be more specific? Did jekyll fail to build the site locally, or was the site built but not displaying correctly?
> 
> > 1. Updated _config.yaml with the remote theme
> > 2. Updated the gemfile
> > 3. Executed "bundle exec Jekyll serve"
> 
> Did you also do `bundle update`? That is always needed after changing the gemfile.
> 
> Try using the [gemfile from CBS-beta](https://github.com/plancomps/CBS-beta/blob/master/Gemfile), which is using this branch.
> 
> If you intend to use the dark color scheme, I recommend using the [combination-rec-nav](https://github.com/pdmosses/just-the-docs/tree/combination-rec-nav) branch instead of rec-nav-2.


The build was successful, but nothing shown up in the screen apart from navigation. I have tried copying the navigation folder to do a sanity check.
![image](https://user-images.githubusercontent.com/29459758/111289070-c97d0080-866a-11eb-89a4-c3b564eaba0f.png)

But things did not turn up as expected in the page.

![image](https://user-images.githubusercontent.com/29459758/111289240-f92c0880-866a-11eb-9ca9-c5e4754e9b59.png)

droy2017 commented:
> 
> 
> > @droy2017 you wrote:
> > > I tried following below steps as recommended above, but it did not work.
> > 
> > 
> > Could you be more specific? Did jekyll fail to build the site locally, or was the site built but not displaying correctly?
> > > 1. Updated _config.yaml with the remote theme
> > > 2. Updated the gemfile
> > > 3. Executed "bundle exec Jekyll serve"
> > 
> > 
> > Did you also do `bundle update`? That is always needed after changing the gemfile.
> > Try using the [gemfile from CBS-beta](https://github.com/plancomps/CBS-beta/blob/master/Gemfile), which is using this branch.
> > If you intend to use the dark color scheme, I recommend using the [combination-rec-nav](https://github.com/pdmosses/just-the-docs/tree/combination-rec-nav) branch instead of rec-nav-2.
> 
> The build was successful, but nothing shown up in the screen apart from navigation. I have tried copying the navigation folder to do a sanity check.
> ![image](https://user-images.githubusercontent.com/29459758/111289070-c97d0080-866a-11eb-89a4-c3b564eaba0f.png)
> 
> But things did not turn up as expected in the page.
> 
> ![image](https://user-images.githubusercontent.com/29459758/111289240-f92c0880-866a-11eb-9ca9-c5e4754e9b59.png)

Made changes to gemfile, but the result it same. Do I need to use collection to make it work? Please note that I am not using collection in this case.

![image](https://user-images.githubusercontent.com/29459758/111291282-0c3fd800-866d-11eb-959a-e39997c8b8de.png)

pdmosses commented:
>  Do I need to use collection to make it work? Please note that I am not using collection in this case.

The page navigation should not require the use of collections.

droy2017 commented:
> 
> 
> > Do I need to use collection to make it work? Please note that I am not using collection in this case.
> 
> The page navigation should not require the use of collections.

Thanks for the confirmation Peter.  Do you have suggestion to make it work?

pdmosses commented:
I've added a simplified test site for multi-level navigation:

https://pdmosses.github.io/test-multi-level/

I suggest to make sure that your site includes all the relevant settings in `_config.yml` and `Gemfile`. It works for me; I hope it will work for you too.

Michael2MacDonald commented:
**_ISSUE!!_**
If you have two pages titled the same thing and one of them is a top-level page, they are both highlighted when you navigate to the not top-level page.
Maybe I have set up the files wrong but I am pretty sure everything is right. My setup only uses the `ancestor` variable and does not use sections.
![image](https://user-images.githubusercontent.com/34051815/118720125-118ce080-b7de-11eb-9012-1b5fb9282a9e.png)

pdmosses commented:
@Michael2MacDonald many thanks for testing this PR, and for your support for adding recursive navigation to the theme!

Thanks also for reporting the nav highlighting issue. Using the same title for a top-level page and an (unrelated) child page isn't excluded by the [assumptions](https://pdmosses.github.io/just-the-docs/tests/navigation/recursion/about#current-proposal) of this PR, but I guess I never tested that case. I'll take a look at how to fix the issue a.s.a.p.

AlanFoster commented:
@pdmosses This looks great, recursive support would be great functionality to add! :+1:

I've been trying to get this running, and I'm unfortunately running into a crash:

```
Liquid Exception: Liquid error [...] Nesting too deep
```

[Replication branch](https://github.com/AlanFoster/just-the-docs-crash)

Running:
```
cd docs
bundle
rm -rf _site && bundle exec jekyll serve --config _config.yml,_config_development.yml
```

I've verified that the repository works against the latest master release, but not the recursive branch. Let me know if there's anything I can do to help :)

pdmosses commented:
@AlanFoster the navigation hierarchy in your [Replication branch](https://github.com/AlanFoster/just-the-docs-crash) has a cycle, due to the duplication of the title `Using Metasploit` in the file [Using-Metasploit.md](https://github.com/AlanFoster/just-the-docs-crash/blob/master/docs/using-metasploit/getting-started/Using-Metasploit.md). A local copy of your site builds OK (using also my `combination-rec-nav` branch) when I change the `title` in the front matter of that file to (e.g.) `Using Metasploit - Getting Started`.

AFAIK, the `Nesting too deep` error is _always_ due to a cycle in the hierarchy. I found this one by using `exclude` to omit folders until Jekyll was able to build the remaining part of the site, then restoring them gradually to locate the cycle. Ideally, the theme would check the navigation hierarchy and report any cycles before proceeding with rendering the pages...

I think the successful build of the same site when using just-the-docs v0.3.3 is due to its implicit distinction of ambiguous titles by their levels. The [Notes](https://pdmosses.github.io/just-the-docs/tests/navigation/recursion/about.html) about recursion mention that recursive navigation assumes the absence of cycles.

AlanFoster commented:
@pdmosses Thanks for taking a look! :+1:

I renamed the file and was able to successfully get the replication branch up and running with `rec-nac-2` 🎉 

Having the requirement of unique titles is turning out to be problematic for myself as I have two `Getting Started` subsections in the docs, one aimed at users, and another for developers:

```
Users
  Getting Started <--
     etc.md

Developers
  Getting Started <--
    etc.md
```

I also ran into a similar issue with having multiple folders named 'Miscellaneous', and it looks like files will appear in the wrong subgrouping. I may run out of synonyms soon, but the functionality is working great! 🏆

pdmosses commented:
> Having the requirement of unique parent titles

That is _not_ a requirement. [As stated in the Notes](https://pdmosses.github.io/just-the-docs/tests/navigation/recursion/about.html#current-proposal): "A child cannot have the same title as its parent or any of its ancestors".

The notes also state:

- If you want the navigation structure in different parts of your website to look the same, you can add the title of the top page of each part as the `ancestor` of all its sub-pages.

- Instead of using `ancestor`, you can set an arbitrary number or string as the `section_id` for each part, and add it as the `in_section` field on all the descendant pages of that part.

The [Sections](https://pdmosses.github.io/just-the-docs/tests/navigation/section/index.html) tests show how to get subsections, to support ambiguous titles within sections.

Let me know if ancestors and sections (or their current implementation) aren't adequate to resolve the ambiguous titles on your site, or if the documentation is unclear.

> I also ran into a similar issue with having multiple distinct folders having generic names such as 'Miscellaneous', and 'Overview', and it looks like files will appear in the wrong subgrouping.

AFAIK, the navigation hierarchy is completely independent of the names of files and folders, so I don't see how that issue can be related to this PR.

emiltin commented:
Hi,
I'm trying your branch `rec-nav-2`(used n this PR). Links in navigation and breadcrumbs are broken because they use https://pdmosses.github.io as the base url.

pdmosses commented:
Hi @emiltin,

Thanks for reporting this. The problem might be caused by my mistaken use of absolute urls instead of relative urls in the nav code (see #543). This issue has been fixed in the `combination-rec-nav` branch (which is the one I use for my own projects). Could you try using that branch instead of `rec-nav-2`?

If that doesn't help, I'll take a closer look, but unfortunately I don't know much about Jekyll configuration and its relation to the theme settings...

emiltin commented:
i will take a look at that branch. but regardless, it seems like a bug to have 'https://pdmosses.github.i' anywhere in the branch?

emiltin commented:
your branch 'combination-nav-rec' also sets the url to your domain:

https://github.com/pmarsceill/just-the-docs/compare/master...pdmosses:combination-rec-nav#diff-ecec67b0e1d7e17a83587c6d27b6baaaa133f42482b07bd3685c77f34b62d883R24

and it has the comment 'TO BE REMOVED'.

AlanFoster commented:
@emiltin If you're referring to the link at the top; and this line of code:
https://github.com/pdmosses/just-the-docs/blob/429b51844fbcd9870dab55c4e844aac302170123/_config.yml#L21-L26

I documented the solution I've used here: https://github.com/pmarsceill/just-the-docs/pull/578#issuecomment-916305038

emiltin commented:
@AlanFoster yes it's that part. thanks for work-around

peterchenadded commented:
@pdmosses did you get a chance to look at issue reported by @Michael2MacDonald 

I had the same issue and had to update below code to be startswith instead of contains. Note liquid had no startswith operator so had to use combination of size, slice and == as an alternative.

https://github.com/pdmosses/just-the-docs/blob/combination-rec-nav/_includes/nav/links.html#L50

```
{%- if nav_page_path contains nav_path -%}
```

pdmosses commented:
@peterchenadded many thanks for trying to use the combination-rec-nav branch, and for the issue reports.

I'm currently giving priority to getting the combination branch (PR #578) ready for merging into the forthcoming v0.4.0 release. After that, I intend to rebase rec-nav-2 on v0.4.0, and resume work on improving it. 

> @pdmosses did you get a chance to look at issue reported by @Michael2MacDonald
> 
> I had the same issue and had to update below code to be startswith instead of contains. Note liquid had no startswith operator so had to use combination of size, slice and == as an alternative.
> 
> https://github.com/pdmosses/just-the-docs/blob/combination-rec-nav/_includes/nav/links.html#L50
> 
>   {%- if nav_page_path contains nav_path -%}

Right, that condition can hold when paths of different lengths have a title in common. Annoying that Liquid doesn't have startswith. Perhaps prefixing both paths by some unused character (e.g., TAB) would eliminate the false positives that arise with contains?

pdmosses commented:
@peterchenadded 

> @pdmosses also any ideas to improve the build times of combination-rec-nav branch?
> 
> For roughly 300 pages it was taking 600 seconds to build for us. Due to below code:
> 
> https://github.com/pdmosses/just-the-docs/blob/combination-rec-nav/_includes/nav/links.html#L56
> 
>   {%- if nav_active or site.nav_cache == false or node.parent -%}
> 
> I removed the `or node.parent` to get the build time down to 200 seconds. Most of our pages are at the bottom of a 4 level tree, so that flag would always evaluate to true and not use the cache include.

I used various heuristics in my efforts to optimise the build time, but they were only tested on my CBS-beta site. I'll take another look at them when I resume work on this PR.

> I still feel even with that improvement build is slow, with the nav taking up most of the time. Is it possible for example to generate the nav only once and some how include the active classes as required for the page?

In a comment at https://github.com/pdmosses/just-the-docs/blob/429b51844fbcd9870dab55c4e844aac302170123/_includes/nav/collection.html#L33 I wrote:

> The presence of navigation expanders requires the navigation links for the entire site to be output on every page

I could add an option to omit nav expanders entirely – users would then need to visit pages in order to see links to their children. Only the nav links for children of the current page, its siblings, and siblings of its ancestors would need to be output. That could make a huge (exponential) difference to page size as well as build time.

> Maybe put the nav items into a dict by page url only once and then use that to generate the page nav html? The specific page url segments can be quickly looked up and add the active classes.

That sounds feasible – but perhaps the currently used caching could be tweaked to give the same effect?

peterchenadded commented:
> @pdmosses 
> Right, that condition can hold when paths of different lengths have a title in common. Annoying that Liquid doesn't have startswith. Perhaps prefixing both paths by some unused character (e.g., TAB) would eliminate the false positives that arise with contains?

Yes, your right that's another approach.

>> @peterchenadded 
>> Maybe put the nav items into a dict by page url only once and then use that to generate the page nav html? The specific >>page url segments can be quickly looked up and add the active classes.

> @pdmosses 
> That sounds feasible – but perhaps the currently used caching could be tweaked to give the same effect?

Yes but need some way to retrieve the dict and then change that data. Do not think include_cached supports that scenario. Current implementation passes the path of the page to the include which makes it difficult to cache.

peterchenadded commented:
@pdmosses maybe the approach to speed up the build is to introduce some JS to added the active classes.

The nav html would always be the same with IDs added to each link. At the end of the nav html we add some JS which use document.getelementbyid to add active classes. This is would be different per page but would be short and easy to generate.

adfoster-r7 commented:
Hi there 👋 

I've been using this pull request's commits over at https://docs.metasploit.com/ for a while now, and the functionality is pretty awesome. Is there anything else needed for this to land? 👀 

As an extra data point for the performance discussion, it takes ~45 seconds to build the 151 pages on our site

pdmosses commented:
Hi @adfoster-r7.

Thanks for the link to your site using the multi-level navigation proposed in this PR, and for the performance data point.

When v0.4.0 has been released, I intend to give updating this PR high priority. Resolving the conflicts should be straightforward. I'll move the multi-level tests to a separate repo (to simplify reviewing).

I'd also like to tweak the multi-level nav to make it backwards compatible with v0.4.0. That may not be so easy…

adfoster-r7 commented:
Awesome! This would be cool functionality to get into the main release 😄 

Just to share some ideas I made on a fork of this work; I updated the navigation to open by default over in this [commit](https://github.com/rapid7/just-the-docs/commit/71772286f92421332a29c75765e7af86402db581). I'm not sure if that's a configuration item that would make sense to add in this PR or not, but thought I'd mention it :+1:

I also added some visual guidelines and a slightly wider navigation bar in this [commit](https://github.com/rapid7/just-the-docs/commit/de685cdfb62d1f962e01898f31c9854ef5ebca0e) to help improve the readability of deeply nested menu items:

<img width="343" alt="image" src="https://user-images.githubusercontent.com/60357436/206936021-f358ee7b-d945-4978-934e-47ccee6dcbc5.png">

mattxwang commented:
Just to chime in - been a bit MIA recently since I'm still recovering from COVID + just wrapping up finals now. Hopefully will be able to get quite a bit of JtD work done in the next month as I'm on holiday, and this is certainly high up on the list!

(going to do my best to get to other issues/PRs for the next few days)

adfoster-r7 commented:
I haven't investigated the specific performance bottlenecks; but I wondered if the performance could be improved with a custom opt-in `_plugin`. i.e. the slow parts could be handled directly in Ruby instead of liquid. It would have to be opt-in functionality though if the plugin worked - as I believe by default plugins aren't supported on Github's default gh-pages, but they're available in Github action, folk's own CI pipelines, or local dev environments

pdmosses commented:
@adfoster-r7 thanks for the suggestion. However, I'm not a Ruby programmer, so I wouldn't develop such a plugin myself.

I'd like to update this PR to use v0.4, then start investigating whether it still suffers from the reported performance issues. But that will have to wait until I get time for it (hopefully from late April).

mattxwang commented:
Just circling back - slowly looking to devote more time on bigger JtD projects (instead of just maintenance). Anything I can do to help get this PR up to speed?

pdmosses commented:
@mattxwang
> Anything I can do to help get this PR up to speed?

I think it would be better to open a _new_ PR for supporting multi-level navigation, rather than continuing this one. The conversation here has become too long for comfort. Having to scroll down through a lot of (mostly resolved) comments would surely discourage discussion and reviews.

Moreover, some sites may have been using this PR branch for the past year or two. To avoid the risk of breaking those sites, I suggest to leave the PR branch as it is now. We can also leave the PR itself open until a new PR for multi-level navigation has been approved and merged.

@peterchenadded and @adfoster-r7 have made some suggestions for reducing the build times for sites using my current implementation of multi-level navigation (which can be painfully long, even when using Jekyll 4). Version 0.4 of JtD significantly reduced the build times for some larger 3-level sites; it's conceivable that incorporating similar optimisations would also help  with multi-level sites.

However, we might also consider a more radical redesign of how we _specify_ navigation in JtD sites. I recently took a closer look at [Hugo], which seems to be generally regarded as the fastest SSG currently available. My curiosity had been piqued by some (partly misleading) comments about slow builds with JtD [on Jekyll Talk].

I also noticed a Hugo theme called [Docsy], which supports several features that we have yet to implement in JtD – including multi-lingual sites, versioned docs, and multi-level navigation! Using the folder structure of the source files to specify the navigation structure (apart from the relative order of pages at the same level) might allow JtD to reduce multi-level build times significantly, as well as circumventing the current need of the present PR for specifying `section` or `ancestor` to disambiguate `parent` title resolution.

So instead of trying to "get this PR up to speed", I'd prefer us to open a fresh _issue_ (or simply a new discussion) to decide how _best_ to support multi-level navigation in JtD – paying special attention to build times – before opening a PR to implement that support. Would that be okay?

[Hugo]: https://gohugo.io
[Docsy]: https://www.docsy.dev
[on Jekyll Talk]: https://talk.jekyllrb.com/t/13-seconds-build-time-per-page/8007

adfoster-r7 commented:
@pdmosses Whatever you think is the best approach to move forwards with things works for me - I'm happy to test any new structure/format for bugs or performance comparisons as well :+1:

Just as an extra datapoint; docusaurus also relies on the folder structure without needing extra frontmatter to be present when generating its sites:

- Example page: https://docusaurus.io/docs/advanced/routing
- Frontmatter: https://github.com/facebook/docusaurus/blob/41a52161fd3df5433ec2ed5e013743cec80a2286/website/docs/advanced/ssg.mdx

They do provide an escape hatch for specifying a [slug](https://github.com/facebook/docusaurus/blob/main/website/docs/advanced/routing.mdx?plain=1#L59) to ignore the folder and generate a custom path though [example](https://github.com/facebook/docusaurus/blob/41a52161fd3df5433ec2ed5e013743cec80a2286/website/docs/migration/migration-manual.mdx?plain=1#L1-L4)

peterchenadded commented:
If we are talking about performance, we need to also assess the underlying technology being used. 

For example, are people actually using liquid in the markdown files?

If we instead use markdown-it an in browser js markdown converter and change to single page design, the pages will load much much faster.

We had actually gone down this path already and the build and page load performance is light and day for our growing 2.5k page site.

adfoster-r7 commented:
> are people actually using liquid in the markdown files?

I can't speak for the majority; For our docs we definitely use liquid for generating the links to other pages ([example](https://github.com/just-the-docs/just-the-docs/blob/d107faaff4bda596e0875b0f5da2e8d1608b9c8f/docs/navigation-structure.md?plain=1#L242). I imagine folk are using different [liquid filters](https://jekyllrb.com/docs/liquid/filters/). For our docs we also register a custom liquid filter to use.

> If we instead use markdown-it an in browser js markdown converter and change to single page design, the pages will load much much faster.

I believe that might impact SEO and search engine indexing, as well as _potentially_ screen readers. In my experience markdown parsers all have different semantics too, so the in-browser rendering might behave differently to Jekyll's parser - which would introduce regressions in behavior

I imagine that could probably be added on as a separate feature from multi-level navigation support either way, but I'll defer to the experts here

peterchenadded commented:
> I can't speak for the majority; For our docs we definitely use liquid for generating the links to other pages (example. I imagine folk are using different liquid filters. For our docs we also register a custom liquid filter to use.

I think we can divide parts of the page that need liquid and parts that do not need it. There is no way to remove liquid for layout files which includes for example header, nav, footer and aux links you mentioned. The actual content shown probably does not need it and pure markdown syntax is sufficient.

> I believe that might impact SEO and search engine indexing, as well as potentially screen readers. In my experience markdown parsers all have different semantics too, so the in-browser rendering might behave differently to Jekyll's parser - which would introduce regressions in behavior

True, but for us our site is private and SEO was not important. Though, we still have the full.md files and should still work with SEO, if we really wanted it to.

Your right about the rendering being different. But actually markdown-it has good number of extensions. So for example if you wanted number lines for your code, you just use the extension.

Agree there will be regressions and some features will change but if build speed and page load performance is of concern, moving to SPA and not using liquid where it is not needed should be considered.

> I imagine that could probably be added on as a separate feature from multi-level navigation support either way, but I'll defer to the experts here

Yes, nav is a different area and definitely needs to be cached for bigger sites. Just following the structure of the file system was good enough for us, but many will probably want front matter to define different titles. It was the worse performing when profiling but after it is fixed it is almost mandatory to fix the other two issues.

Forgot to mention search, which makes the page unresponsive for a second as it tries to load in search-data.json on every page you visit. We actually had to setTimeout and chunk the loading of that file to make the UI responsive.

mattxwang commented:
Hi all, thanks for the lively discussion! Bigger picture, I think a new issue could be appropriate; I could also put some work in simply porting this PR over to `main` (on a new branch, to avoid disrupting this one).

Let me address a few quick conversation topics:

- personally, I think rendering markdown client-side is significantly out of scope for this project. There are other solutions that work quite well already, and in those cases, you probably shouldn't be using Jekyll; it is intentionally a static-site generator. The "hybrid" approach that's being alluded to - code splitting and doing the equivalent of SSR/islands for SSG - is way out of scope of most Jekyll themes. I think that would be more appropriate for the modern JS stack for this, e.g. Astro/Next/11ty and associated tooling.
- file-based navigation is certainly appealing! I will point out that we are somewhat limited by Jekyll itself and that our theme is intended to work out-of-the-box with GitHub Pages (not necessarily a bad thing); since file-based navigation is *not* first-class, we'd have to do some hackiness to get it to work. The tradeoff is that Hugo is arguably less portable (depends on who you ask). Docusaurus is becoming the gold standard for OSS docs (Stylellnt has been using it for a while, for example), so it definitely makes sense to look at what they're doing.
- I do think an opt-in Ruby plugin is within the cards; pretty sure that we can have users opt-in with a `_config.yml` parameter and only load in the plugin/file at runtime (i.e. when compiling). This preserves backwards compatibility.

Of course, these are all my opinions, and I'm happy to hear out some other perspectives!

In terms of my personal roadmap as a maintainer, there are some prerequisites I'll do *before* I work on this:

- merging in PRs #1169 and #1236
- guiding in automatic theme switching (many open PRs for this, or I'll write my own)

After that, I'm happy to then dedicate my maintenance time to pushing this PR forward.

In the meantime, if someone would either like to:

- fork this branch and open an up-to-date PR
- or, experiment with any of the above suggestions (with the caveat that I may deem an approach out of the spirit of JtD)

feel free to do so - happy to review!

pdmosses commented:
>  Bigger picture, I think a new issue could be appropriate; I could also put some work in simply porting this PR over to `main` (on a new branch, to avoid disrupting this one).

In any case, I think I first should move all the multi-level tests from this PR branch to (a branch of) `just-the-doc-tests`.

BTW, I'm currently using the [combination-rec-nav branch](https://github.com/pdmosses/just-the-docs/tree/combination-rec-nav) of my fork of JtD to publish a draft proposal for [multi-level navigation docs](https://pdmosses.github.io/just-the-docs/docs/navigation-structure/) and some [accompanying notes](https://pdmosses.github.io/just-the-docs/tests/navigation/recursion/about). Netlify provides deploy previews of all PRs, so I won't need to change that branch.

> Let me address a few quick conversation topics:
> 
> * personally, I think rendering markdown client-side is significantly out of scope for this project. 

I know nothing about it, and I wouldn't be able to contribute.

> * file-based navigation is certainly appealing! I will point out that we are somewhat limited by Jekyll itself and that our theme is intended to work out-of-the-box with GitHub Pages (not necessarily a bad thing); since file-based navigation is _not_ first-class, we'd have to do some hackiness to get it to work. The tradeoff is that Hugo is arguably less portable (depends on who you ask). Docusaurus is becoming the gold standard for OSS docs (Stylellnt has been using it for a while, for example), so it definitely makes sense to look at what they're doing.

I've also looked at [Material for MkDocs](https://squidfunk.github.io/mkdocs-material/), which has already implemented many of the features that we intend to support in JtD (multi-level navigation, i18n, versioned docs, toggled between color schemes, …). It seems that the navigation hierarchy there can either be left implicit in the directory structure, using file names as titles – with optional front matter to override the URL and/or the title displayed in the navigation panel – or specified in a single YAML file that associates titles with file names in the intended order and nesting. Jekyll provides variables such as `page.path`, and I don't see why any hacking would be needed to implement the same idea in JtD. There's also [Docsy Jekyll](https://vsoch.github.io/docsy-jekyll/), a port of the Hugo theme [Docsy](https://www.docsy.dev/). 

> * I do think an opt-in Ruby plugin is within the cards; pretty sure that we can have users opt-in with a `_config.yml` parameter and only load in the plugin/file at runtime (i.e. when compiling). This preserves backwards compatibility.

I've never programmed in Ruby, and I wouldn't be able to contribute to developing such a plugin. If we'd need to maintain consistency between a plugin and a Liquid implementation of multi-level navigation, some other maintainers would need to take responsibility for that.

> Of course, these are all my opinions, and I'm happy to hear out some other perspectives!
> 
> In terms of my personal roadmap as a maintainer, there are some prerequisites I'll do _before_ I work on this:
> 
> * merging in PRs [Use `rem` instead of `px` for all text related CSS properties #1169](https://github.com/just-the-docs/just-the-docs/pull/1169) and [Add visual tests with Storybook & Chromatic #1236](https://github.com/just-the-docs/just-the-docs/pull/1236)
> * guiding in automatic theme switching (many open PRs for this, or I'll write my own)
> 
> After that, I'm happy to then dedicate my maintenance time to pushing this PR forward.

I have a pressing conference submission deadline at the end of June. After that, I should have time to participate in further discussion and potential implementation of multi-level navigation.

> In the meantime, if someone would either like to:
> 
> * fork this branch and open an up-to-date PR

I think we need an in-depth discussion of _what_ we would like to implement _before_ opening a new PR on this topic.

There's an open discussion #425 with the question "Navigation levels limitation?". My response is now outdated, but I think it would be easy enough to update it, as a starting point for a fresh discussion of our current ideas for multi-level navigation. 

An alternative might be to revisit the open issue #219 on "Navigation doesn't work with grand-grand-children". However, the linear structure of an issue doesn't seem as convenient as a discussion with threads of comments on potential answers, and discussions have the possibility of sorting them by the date of latest activity.

ragalgut commented:
It would be fantastic if this went ahead. A recursive navigation.

pdmosses commented:
@ragalgut Thanks for your comment!

PR #1397 should make it much easier to add recursive navigation (by reducing dependence on the `parent`, `grand_parent`, and `has_children` relations).

I expect to have time _this_ month to investigate whether the directory structure could determine the navigation hierarchy (apart from the `title` and `nav_order`, which would still be specified in front-matter). I think that would have some significant benefits, although there might also be some drawbacks.

I suggest to continue the discussion in #1277, to make it more visible.

didacrios commented:
Hi @pdmosses is there any consensus achieved related to this point? Seen the branch from the proposal far from the current main branch of the project and conflicts raised related to this https://just-the-docs.com/migration/#moved-include

pdmosses commented:
This PR is to be superseded by #1431.

Users with websites built with the PR branch `rec-nav-2` should consider migrating to the `multi-level` branch, as `rec-nav-2` will not be updated, and may eventually be deleted.

pdmosses commented:
> Hi @pdmosses is there any consensus achieved related to this point?

@didacrios PR #1431 generalises the latest JtD release to allow any number of navigation levels. However, we cannot claim that a consensus has been achieved, as there has not been any further discussion of the adopted approach.

pdmosses commented:
Superseded by #1431
